src/plot_api/plot_api.js

```diff
@@ -1528,57 +1430,14 @@ function _restyle(gd, aobj, _traces) {
 
             if(newVal === undefined) continue;
 
-            // setting bin or z settings should turn off auto
-            // and setting auto should save bin or z settings
-            if(zscl.indexOf(ai) !== -1) {
-                doextra('zauto', false, i);
-            }
-            if(cscl.indexOf(ai) !== -1) {
-                doextra('cauto', false, i);
-            }
-            else if(ai === 'colorscale') {
-                doextra('autocolorscale', false, i);
-            }
-            else if(ai === 'autocolorscale') {
-                doextra('colorscale', undefined, i);
-            }
-            else if(ai === 'marker.colorscale') {
-                doextra('marker.autocolorscale', false, i);
-            }
-            else if(ai === 'marker.autocolorscale') {
-                doextra('marker.colorscale', undefined, i);
-            }
-            else if(ai === 'zauto') {
-                doextra(zscl, undefined, i);
-            }
-            else if(xbins.indexOf(ai) !== -1) {
-                doextra('autobinx', false, i);
-            }
-            else if(ai === 'autobinx') {
-                doextra(xbins, undefined, i);
-            }
-            else if(ybins.indexOf(ai) !== -1) {
-                doextra('autobiny', false, i);
-            }
-            else if(ai === 'autobiny') {
-                doextra(ybins, undefined, i);
-            }
-            else if(contourAttrs.indexOf(ai) !== -1) {
-                doextra('autocontour', false, i);
-            }
-            else if(ai === 'autocontour') {
-                doextra(contourAttrs, undefined, i);
-            }
-            // heatmaps: setting x0 or dx, y0 or dy,
-            // should turn xtype/ytype to 'scaled' if 'array'
-            else if(['x0', 'dx'].indexOf(ai) !== -1 &&
-                    contFull.x && contFull.xtype !== 'scaled') {
-                doextra('xtype', 'scaled', i);
-            }
-            else if(['y0', 'dy'].indexOf(ai) !== -1 &&
-                    contFull.y && contFull.ytype !== 'scaled') {
-                doextra('ytype', 'scaled', i);
+            valObject = PlotSchema.getTraceValObject(contFull, param.parts);
+
+            if(valObject && valObject.impliedEdits && newVal !== null) {
+                for(var impliedKey in valObject.impliedEdits) {
+                    doextra(Lib.relativeAttr(ai, impliedKey), valObject.impliedEdits[impliedKey], i);
```

Comment:
One question about implied edits is whether it moves away from declarative plots. That is, if you set `foo` (which, e.g., implicitly also sets `bar: false`), is it valid to then set `bar: true`? Is there pathway that might accidentally circumvent the implied edits and lead to a bit of inconsistency? Would it be better/worse to think of  implied edits as a concern of the plotly API consumer rather than the plotly API itself? Like the workspace would recognize the attribute you're modifying has an implied edit and send *both* of those to the API? It seems nice from an end-user perspective; my only concern is that it might be surprising/confusing to modify an attribute and find that additional changes have been made to your plot state.

Reply:
This PR by itself isn't moving toward or away from anything as far as users are concerned... all of these `impliedEdits` were previously handled by dedicated blocks in `_restyle`/`_relayout`. Without them, a call like `relayout(gd, 'xaxis.range', [5, 10])` wouldn't do anything if `xaxis.autorange` was true.

Looking ahead though to `Plotly.react`, I'm hoping that by having `impliedEdits` in the schema we can help people make these implied changes in a transparent way - so if you have a `range` control that modifies the state and passes it back to the plot in `Plotly.react`, it would be able to look up the implied changes we think it should make, but the control would have final say over whether to apply those changes or not.

Reply:
Got it. Thanks for the clarification.